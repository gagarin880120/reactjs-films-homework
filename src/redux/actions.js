function searchResultsAction(results) {
  return {
      type: 'SEARCH',
      results: results
  };
}

function genresAction(genres) {
  return {
      type: 'GENRES',
      genres: genres
  };
}

export { searchResultsAction, genresAction };
