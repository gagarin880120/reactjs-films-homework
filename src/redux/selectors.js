import getGenresString from '../helpers/helpers';

const searchResultsSelector = (state) => {
  state.searchResults.map((v) => {
    const obj = v;
    obj.genres = getGenresString(v.genre_ids, state.genres);
    return { ...obj };
  });
  return state.searchResults;
};

const modalSelector = (state) => state.isModalOpen;

const trailerSelector = (state) => state.trailerURL;


export { searchResultsSelector, modalSelector, trailerSelector };
