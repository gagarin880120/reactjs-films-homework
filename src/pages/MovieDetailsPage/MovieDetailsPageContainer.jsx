import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import MovieDetailsPage from './MovieDetailsPage';
import Spinner from '../../components/Spinner/Spinner';
import {
  areMoviesLoadedSelector, resultsSelector, isMovieLoadedSelector,
} from '../../redux/selectors';
import { getMovies, setCurrentAPIRequest } from '../../redux/actions';
import styles from './MovieDetailsPage.module.scss';

export function MovieDetailsPageContainer({
  areMoviesLoaded, results, isMovieLoaded, onPageLoad,
}) {
  return (
    isMovieLoaded
      ? (
        <MovieDetailsPage
          areMoviesLoaded={areMoviesLoaded}
          results={results}
          onPageLoad={onPageLoad}
        />
      ) : (
        <div className={styles.pageLoading}>
          <Spinner />
        </div>
      )
  );
}

export const mapStateToProps = (state) => ({
  areMoviesLoaded: areMoviesLoadedSelector(state),
  results: resultsSelector(state),
  isMovieLoaded: isMovieLoadedSelector(state),
});

export const mapDispatchToProps = (dispatch) => ({
  onPageLoad() {
    dispatch(setCurrentAPIRequest('popular'));
    dispatch(getMovies('popular'));
  },
});

MovieDetailsPageContainer.propTypes = {
  areMoviesLoaded: PropTypes.bool,
  results: PropTypes.instanceOf(Array),
  isMovieLoaded: PropTypes.bool,
  onPageLoad: PropTypes.func,
};

MovieDetailsPageContainer.defaultProps = {
  areMoviesLoaded: false,
  results: [],
  isMovieLoaded: false,
  onPageLoad: null,
};

export default connect(mapStateToProps, mapDispatchToProps)(MovieDetailsPageContainer);
