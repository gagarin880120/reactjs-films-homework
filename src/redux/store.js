import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';

const initialState = {
  searchResults: [],
  genres: [],
  isModalOpen: false,
  trailerURL: '',
  isTrailerLoaded: false,
};

export function reducer(state = initialState, action) {
  switch (action.type) {
    case 'SEARCH':
      return { ...state, searchResults: action.results };
    case 'GENRES':
      return { ...state, genres: action.genres };
    case 'MODAL':
      return { ...state, isModalOpen: action.isModalOpen };
    case 'TRAILER':
      return { ...state, trailerURL: action.trailerURL };
    case 'IS_TRAILER_LOADED':
      return { ...state, isTrailerLoaded: action.isTrailerLoaded };
    default:
      return state;
  }
}

export const store = createStore(reducer, applyMiddleware(thunk));
