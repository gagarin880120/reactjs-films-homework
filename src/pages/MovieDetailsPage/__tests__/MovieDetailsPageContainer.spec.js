import React from 'react';
import { MovieDetailsPageContainer, mapStateToProps } from '../MovieDetailsPageContainer';
import ShallowRenderer from 'react-test-renderer/shallow';
import { setModal, getTrailer } from '../../../redux/actions';

jest.mock('../../../redux/actions',() => ({
  setModal: jest.fn().mockReturnValue(false),
  getTrailer: jest.fn().mockReturnValue('trailer'),
}));

describe('MovieDetailsPageContainer', () => {
  describe('component', () => {
    test('should render without crashing', () => {
      const renderer = new ShallowRenderer();
      renderer.render(<MovieDetailsPageContainer />);
      const result = renderer.getRenderOutput();
      expect(result).toMatchSnapshot();
    });

    test('should render spinner if movie is not loaded', () => {
      const renderer = new ShallowRenderer();
      renderer.render(<MovieDetailsPageContainer isMovieLoaded={true} />);
      const result = renderer.getRenderOutput();
      expect(result).toMatchSnapshot();
    });
  });


  describe('mapStateToProps', () => {
    test('should return the right value', () => {

      const initialState = {
        areMoviesLoaded: true,
        results: [{title:"Star Wars", id:15, genre_ids:[28]}],
        genres: [{id: 28, name: 'Action'}],
        movieDetails: {title: 'Matrix', runtime: 120, vote_average: 8.6, genres: [{name: 'Adventure'}]},
      };
      expect(mapStateToProps(initialState).results).toEqual([
        {title:"Star Wars", id:15, genre_ids:[28], genres: 'Action'}
      ]);
      expect(mapStateToProps(initialState).movie).toEqual({
        genres: 'Adventure',
        runtime: '2h',
        stars: [
          {
            id: 0,
            type: 'full',
          },
          {
            id: 1,
            type: 'full',
          },
          {
            id: 2,
            type: 'full',
          },
          {
            id: 3,
            type: 'full',
          },
          {
            id: 4,
            type: 'half',
          },
        ],
        title: 'Matrix',
        vote_average: 8.6,
      });
    });
  });

});
