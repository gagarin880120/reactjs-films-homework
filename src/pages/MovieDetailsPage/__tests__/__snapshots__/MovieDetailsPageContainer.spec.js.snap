// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieDetailsPageContainer component should render spinner if movie is not loaded 1`] = `
<MovieDetailsPage
  areMoviesLoaded={false}
  movie={Object {}}
  results={Array []}
/>
`;

exports[`MovieDetailsPageContainer component should render without crashing 1`] = `
<div
  className="pageLoading"
>
  <Spinner />
</div>
`;
